### PR TITLE
[oneDNN] Enable Softmax with oneDNN library

### PR DESCRIPTION
**Pull request by mdfaijul**
This PR enables softmax forward op using oneDNN library. It uses in-place computation whenever possible. It improves performance of some models wherein the softmax takes up significant time.

The following are performance data on some micro-benchmarks. The performance data were collected on Intel Xeon CPUs using Eigen ThreadPool.

## Microbenchmarks

Tensor Dims | Numeric Type | Intra-op Threads | HW | oneDNN (ns) | Eigen (ns) | Speedup
-- | -- | -- | -- | -- | -- | --
1x16x384x384 | FLOAT32 | 4 | Xeon 28 core | 435505 | 849837 | 1.95x
16x16x384x384 | FLOAT32 | 4 | Xeon 28 core | 30608250 | 39503349 | 1.29x
32x1008 | FLOAT32 | 1 | Xeon 28 core | 28455 | 33078 | 1.16x
128x1008 | FLOAT32 | 1 | Xeon 28 core | 83065 | 123516 | 1.49x
32x1008 | FLOAT32 | 4 | Xeon 28 core | 30585 | 32165 | 1.05x
128x1008 | FLOAT32 | 4 | Xeon 28 core | 41348 | 127743 | 3.09x
1x16x384x384 | BFLOAT16 | 4 | Xeon 26 core | 430448 | 867489 | 2.02x
16x16x384x384 | BFLOAT16 | 4 | Xeon 26 core | 17430198 | 25598959 | 1.47x
32x1008 | BFLOAT16 | 1 | Xeon 26 core | 31498 | 49853 | 1.58x
128x1008 | BFLOAT16 | 1 | Xeon 26 core | 91812 | 159045 | 1.73x
32x1008 | BFLOAT16 | 4 | Xeon 26 core | 29836 | 49224 | 1.65x
128x1008 | BFLOAT16 | 4 | Xeon 26 core | 43047 | 122662 | 2.85x

This PR improves inference performance by 12% on some models that use softmax.

### CODE REVIEW

**Comment by mdfaijul** [comment]
@penpornk When I submitted this PR, github automatically assigned you as reviewer. Anyway, may I request for a review.

**Comment by gbaned** [comment]
@penpornk Can you please review this PR ? Thank you!

**Comment by ashiqimranintel** [comment]
Any update on this PR?

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by mdfaijul** [comment]
@penpornk Any update on this PR?

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+BM_oneDNN_Softmax(float, 4, "float32_ImageNet_batch_size_32", 32, 1008);
+BM_oneDNN_Softmax(float, 4, "float32_ImageNet_batch_size_128", 128, 1008);
+BM_oneDNN_Softmax(bfloat16, 4, "bfloat16_BERT_batch_size_1", 1, 16, 384, 384);
+BM_oneDNN_Softmax(bfloat16, 4, "bfloat16_BERT_batch_size_16", 16, 16, 384, 384);
+BM_oneDNN_Softmax(bfloat16, 1, "bfloat16_ImageNet_batch_size_32", 32, 1008);
+BM_oneDNN_Softmax(bfloat16, 1, "bfloat16_ImageNet_batch_size_128", 128, 1008);
+BM_oneDNN_Softmax(bfloat16, 4, "bfloat16_ImageNet_batch_size_32", 32, 1008);
+BM_oneDNN_Softmax(bfloat16, 4, "bfloat16_ImageNet_batch_size_128", 128, 1008);
+
+// Benchamark with Eigen library.
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc:143]: Nit: Typo
```suggestion
// Benchmark with Eigen library.
```

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+BM_Eigen_Softmax(float, 1, "float32_ImageNet_batch_size_32", 32, 1008);
+BM_Eigen_Softmax(float, 1, "float32_ImageNet_batch_size_128", 128, 1008);
+BM_Eigen_Softmax(float, 4, "float32_ImageNet_batch_size_32", 32, 1008);
+BM_Eigen_Softmax(float, 4, "float32_ImageNet_batch_size_128", 128, 1008);
+BM_Eigen_Softmax(bfloat16, 4, "bfloat16_BERT_batch_size_1", 1, 16, 384, 384);
+BM_Eigen_Softmax(bfloat16, 4, "bfloat16_BERT_batch_size_16", 16, 16, 384, 384);
+BM_Eigen_Softmax(bfloat16, 1, "bfloat16_ImageNet_batch_size_32", 32, 1008);
+BM_Eigen_Softmax(bfloat16, 1, "bfloat16_ImageNet_batch_size_128", 128, 1008);
+BM_Eigen_Softmax(bfloat16, 4, "bfloat16_ImageNet_batch_size_32", 32, 1008);
+BM_Eigen_Softmax(bfloat16, 4, "bfloat16_ImageNet_batch_size_128", 128, 1008);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc:155]: Nit: Please add an empty line after this line.

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+}  // namespace
+}  // namespace tensorflow
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc:157]: Nit: Please add an empty line after this line.

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+      ->set_opt_level(OptimizerOptions::L0);
+  test::Benchmark("cpu", g, &opts, nullptr, nullptr, "",
+                  /*old_benchmark_api*/ false)
+      .Run(state);
+  state.SetItemsProcessed(shape.num_elements() * state.iterations());
+  state.SetLabel(label);
+}
+
+template <typename T>
+static void BM_Eigen_Softmax(::testing::benchmark::State& state,
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc:56]: This and `BM_oneDNN_Softmax` are very similar, I think we can consolidate them into one function (and add another input parameter `bool onednn` to decide what op should be put in the graph).

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+ * The macro PP_NARG returns the number of arguments that have been passed to
+ * it.  Currrently, it is prepared to support upto 5 arguments. It is straight
+ * forward to extend to a higher number of arguments.
+ */
+#define PP_NARG(...) PP_NARG_(__VA_ARGS__, PP_RSEQ_N())
+#define PP_NARG_(...) PP_ARG_N(__VA_ARGS__)
+#define PP_ARG_N(_1, _2, _3, _4, _5, N, ...) N
+#define PP_RSEQ_N() 5, 4, 3, 2, 1, 0
+
+// For a tensor shape {a, b, c, d} we want to produce a token a_b_c_d
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc:95]: Nit: Adding a comma.
```suggestion
// For a tensor shape {a, b, c, d}, we want to produce a token a_b_c_d
```

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                   JOIN(CONCAT_DIMS, PP_NARG(__VA_ARGS__))(__VA_ARGS__))(      \
+      ::testing::benchmark::State & state) {                                   \
+    BM_Eigen_Softmax<dtype>(state, {__VA_ARGS__}, num_threads, label);         \
+  }                                                                            \
+  WRAP_BENCHMARK(JOIN(BM_Eigen_Softmax_##dtype##_intraop_##num_threads##_dims, \
+                      JOIN(CONCAT_DIMS, PP_NARG(__VA_ARGS__))(__VA_ARGS__)))   \
+      ->UseRealTime()
+
+// Benchmark with oneDNN library.
+BM_oneDNN_Softmax(float, 4, "float32_BERT_batch_size_1", 1, 16, 384, 384);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc:130]: Maybe add another macro that defines both the oneDNN and Eigen benchmark at once? E.g.,
```
#define BM_Softmax(dtype, num_threads, label, ...) \
    BM_oneDNN_Softmax(...); \
    BM_Eigen_Softmax(...); 
```

This way, we don't have to update two places when we want to edit / add shapes.

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+#define BM_oneDNN_Softmax(dtype, num_threads, label, ...)                    \
+  static void JOIN(BM_oneDNN_Softmax_##dtype##_intraop_##num_threads##_dims, \
+                   JOIN(CONCAT_DIMS, PP_NARG(__VA_ARGS__))(__VA_ARGS__))(    \
+      ::testing::benchmark::State & state) {                                 \
+    BM_oneDNN_Softmax<dtype>(state, {__VA_ARGS__}, num_threads, label);      \
+  }                                                                          \
+  WRAP_BENCHMARK(                                                            \
+      JOIN(BM_oneDNN_Softmax_##dtype##_intraop_##num_threads##_dims,         \
+           JOIN(CONCAT_DIMS, PP_NARG(__VA_ARGS__))(__VA_ARGS__)))            \
+      ->UseRealTime()
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc:117]: Nit: We care about total process CPU time more than wall time.
```suggestion
      ->MeasureProcessCPUTime()
```

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+#define BM_Eigen_Softmax(dtype, num_threads, label, ...)                       \
+  static void JOIN(BM_Eigen_Softmax_##dtype##_intraop_##num_threads##_dims,    \
+                   JOIN(CONCAT_DIMS, PP_NARG(__VA_ARGS__))(__VA_ARGS__))(      \
+      ::testing::benchmark::State & state) {                                   \
+    BM_Eigen_Softmax<dtype>(state, {__VA_ARGS__}, num_threads, label);         \
+  }                                                                            \
+  WRAP_BENCHMARK(JOIN(BM_Eigen_Softmax_##dtype##_intraop_##num_threads##_dims, \
+                      JOIN(CONCAT_DIMS, PP_NARG(__VA_ARGS__))(__VA_ARGS__)))   \
+      ->UseRealTime()
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc:127]: Same here.
```suggestion
      ->MeasureProcessCPUTime()
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for the delay!

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc (hunk 3 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc:125]: Done.

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc (hunk 2 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc]: Done.

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc (hunk 1 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc]: Done.

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc (hunk 6 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc]: Done. Thanks for the nice suggestions!

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc (hunk 8 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc]: Done.

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc (hunk 7 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc]: Done.

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc (hunk 5 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc]: Done.

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc (hunk 4 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc]: Done.

**Review by mdfaijul — COMMENTED** [review]
@penpornk Thank you very much for reviewing this PR. I have addressed all the comments. Please check.

**Comment by mdfaijul** [comment]
@penpornk I have addressed the comments. Please check. Thanks!

**Comment by penpornk** [comment]
@mdfaijul Happy new year and sorry for the delay! The PR looks good to me. I just needed to check if it's okay to use a code snippet from the Internet. Will get back to you as soon as I got the answer.

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+ * The following trick for number of variadic macro arguments is taken from:
+ * http://groups.google.com/group/comp.std.c/browse_thread/thread/77ee8c8f92e4a3fb/346fc464319b1ee5?pli=1
+ * The macro PP_NARG returns the number of arguments that have been passed to
+ * it.  Currrently, it is prepared to support upto 5 arguments. It is straight
+ * forward to extend to a higher number of arguments.
+ */
+#define PP_NARG(...) PP_NARG_(__VA_ARGS__, PP_RSEQ_N())
+#define PP_NARG_(...) PP_ARG_N(__VA_ARGS__)
+#define PP_ARG_N(_1, _2, _3, _4, _5, N, ...) N
+#define PP_RSEQ_N() 5, 4, 3, 2, 1, 0
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc:71]: Unfortunately, I don't think we can take this code since it's not published under a license.

Can we explicitly specify the number of dims as a `BM_` macro parameter instead? E.g.,
```c++
#define BM_oneDNN_Softmax(dtype, num_threads, label, num_dims, ...)                    \
  static void JOIN(BM_oneDNN_Softmax_##dtype##_intraop_##num_threads##_dims, \
                   JOIN(CONCAT_DIMS, num_dims)(__VA_ARGS__))(    \
      ::testing::benchmark::State & state) {                                 \
    BM_Softmax<dtype>(state, {__VA_ARGS__}, num_threads, label, true);       \
  }  
  ...

// Example usage:
BM_Softmax(float, 1, "float32_ImageNet_batch_size_32", /*num_dims=*/2, 32, 1008);
```

(The code may not compile because I didn't test it.)

**Review by penpornk — CHANGES_REQUESTED** [review]

**Comment by mdfaijul** [comment]
@penpornk Thank you very much for the suggestion. I have removed the code that is not published under a license. Please check.

File: tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc (hunk 9 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc]: Done.

**Review by penpornk — APPROVED** [review]
Thank you for the change and sorry for the inconvenience!

**Comment by penpornk** [comment]
[ARM CI failures](https://github.com/tensorflow/tensorflow/actions/runs/3883926311/jobs/6626240719) seem to be already failing tests (7 tests; no new failures).
